### ISSUE TITLE
Floppy not appearing under the explorer in W7 and XP with 45C and 46a >= 2014-11-03

### ISSUE THREAD

**Issue by mapibid** [chenall/grub4dos#112]
Hello,

Having switched recently to a fresher version of grub4dos (my previous version dated from 2013),
I am getting a bug of floppy drive not avaiiable under Windows 7 and XP = no A: on CLI nor in windows Explorer, errors saying there is no such drive.

This is only when booting through grub4dos (I could determine when >= 2014-11-03).

When booting through normal Windows MBR (either prevmbr or replacing grub4dos MBR by standard WIN 7 MBR), all is fine.
As soon as I put back grub4dos.45c or .46a in the MBR, and use GRLDR >= 2014-11-03 ... pfuit .. the floppy drives is no more visible nor accessible.

I could trace things back to the fix:
_

> 2014-11-03 (tinybit) mask off non-booting floppy drives in order to gain more stability.

_

It can be found here:
https://github.com/chenall/grub4dos/commit/5a887fc1b0c8947116bb96850cd83af854665c49

This fix writes in the BDA (Bios Data Area), to remove presence of the Floppy when we boot on something else than the floppy ..
No wonder the floppy drive does not appear later in Windows !!

See
  http://wiki.osdev.org/Memory_Map_%28x86%29
  http://files.osdev.org/mirrors/geezer/osd/ram/index.htm#layout

and for details on 0x0410 contents:
  http://stanislavs.org/helppc/bios_data_area.html
  http://www.bioscentral.com/misc/bda.htm

To remove the effect of the fix, and prove it is the culprit, in GRLDR >= 2014-11-03, search for
the instruction:
andb    $0x3F, 0x0410

For example, using Hiew32,
 in GRLDR of grub4dos-0.4.6a-2016-03-04.7z
 search for hex string 80 26 10 04
000022B5: 802610043F                   and         b,[0410],03F ;"?"

Change 3F to FF so that we have:
000022B5: 80261004FF                   and         b,[0410],0FF ;" "

There is another one to fix (next search):
000022E3: 80261004FE                   and         b,[0410],0FE ;"¦"

becomes
000022E3: 80261004FF                   and         b,[0410],0FF ;" "

Save, and replace GRLDR with this patched one ..
.. boot with it ..
.. and we can see again "Floppy Disk Drive (A:)" in the Explorer, and in CLI !!

Can you revert the 2014-11-03 fix, or make things less impacting ?

It is quite annoying to not be able to access the Floppy drive from Windows :-(
so I hope you will agree and be able to do something about it, because patching by hand is not the most convenient thing to do for everybody.

Thank you in advance, and thank you by the way for maintaining this excellent piece of software which is really useful, keep going :-)

**Comment by steve6375** [comment]
What happens if you use
map --floppies=1
in your grub4dos boot menu?

**Comment by steve6375** [comment]
If you read 0x410 in grub4dos, then bits 0 and 6&7 are cleared.
I configured a VM with two floppy disks and these bits were cleared.
However, if you do a map --status command, it does say floppies_orig=2
![image](https://cloud.githubusercontent.com/assets/1469350/14063095/44467fe6-f3b1-11e5-8ad9-7b05642bd5e9.png)

so grub4dos does know the machine originally had 2 floppy drives.

You could change the byte using grub4dos...

read 0x410
#use 0x41 if two floppies
set /A A=%@retval%|0x1
write 0x410 %A%

**Comment by mapibid** [comment]
Hello steve6375, thank you for you answer and suggestions.
Yes I was looking at your site http://www.rmprepusb.com/tutorials/grub4dos/grub4dos-commands-primer to do something like that before patching directly GRLDR, but:

1) I was not sure of the exact syntax to use
2) I was not sure that read and write would do exactly bytes, and if they would not write more, having side effects
3) I am still not sure yet of when grub4dos does that in its code with regards to the.configfile sequence ..
- Does it do the clearing of bits 7-6 and 0 when taking hand from the MBR ?
- Does it do it during a particular command, like chainloader or other ?
- Does it do it on the final "boot" at the end of an entry ?
- Does it do it multiple times ?

Depending on answers to these questions, then setting things with read/write inside the configfile could have no effect at all because occurring before the code clears them.

Also, on your suggestion of map --floppies=1, I thought a little about it, but discarded it initially because I am not hooked, my drive map table is empty for the XP and Win7 title entries, and I understood map with floppies was good for "virtual floppies", but would not have effect on real ones. Maybe wrongly, correct me if so ..

Last, I wanted to prove that the fix was at the origin of the problem, so I decided to patch it directly .. this was the only way to make 100% sure it is at the root of things.

At this stage, I am not truly convinced that any of the above tricks would robustly do something, or maybe not all the time, so I'd rather that there is a correction in the code if this agreeable to grub4dos maintainers.
Besides, my feeiling is that tweaking with the BDA is a little dangerous on side effects, see in this case how it affects later the OS to boot (I didn't check with Linux yet), so I would not favor keeping this thing around.

**Comment by steve6375** [comment]
I am worried that 'correcting' the code will affect existing usage and break things. In particular, setting up F6 floppies for booting XP ISOs. The code was added for a reason (what reason?) and no one uses floppy disks these days - they are totally obsolete!

If you use
read 0x410
map --floppies=2
read 0x410

You will see that the map command immediately changes 410.

It surely would be simple for you to test the suggested code in my previous comment to see if it fixes the problem? I am not saying that the code is a fix, I am just asking you to try it, to see if it works or not because that may help us understand what is going on.
It is classic science - you say 'what if', you design an experiment, you implement the experiment, you collect the results, you interpret the results, you gain knowledge.

**Comment by mapibid** [comment]
Indeed, your reasoning is good, and anyway you have way better knowledge of the map command and other things in grub4dos than me, that is why I was saying "correct me if so .."   :-)

So , I tried the experience and I can confirm that adding
   map --floppies=1
just before "boot" is a working workaround for XP and Win7.

No fix of course from my point of view, because I find it quite odd to have to say to the system how many floppies it has in entries of the configfile ... and I feel it should be the other way around = when I want no floppies to interfere for reason X or Y, then specify  map --floppies=0   on purpose.

Thanks for your suggestion anyway, so we now have two workarounds on the problem:
- patch the code
- use map --floppies=X with X being the number of floppies in the system.

At this stage, if tinybit (if he is still around ?) or yaya could provide context / answers, that would help, because the code was added for a reason, but we don't know which one.
Having history of the "why", would be good to understand if there is no better mean than the current one - which has at least this side effect on the booted OS (and who knows what else ?) - and if that is a necessary evil.

Cheers.

**Comment by steve6375** [comment]
I agree with you, but I would like to understand what problem this causes you. Are you saying you need to access a floppy disk? i.e. is it actually a problem that affects you or anyone?
It must be a very old system you have and most people use a USB FDD if they need to access an old floppy disk...

**Comment by mapibid** [comment]
Yes, I am using my floppy drive to generate for example floppies with latest AMD AHCI / RAID (my current motherboard) drivers for Win7 repair boot from the Win7 CD (load driver option). Or also to regenerate boot CD's with a DOS system on it and latest partguru DOS inside it .. etc ..

My system is not old, that is just a case with floppy drive, CD/DVD drives, disk drives in which I put my successive motherboards when I replace them.

**Comment by steve6375** [comment]
It would be nice to know why the code fix was done.
I would think that a possible fix would be for grub4dos to restore the floppies_orig number before a 'boot' is executed - but only if no floppy mapping has been done by grub4dos before it loads an OS.

**Comment by mapibid** [comment]
Agree with you. Let's wait for one of the maintainers advice then :-)
